src/promql/src/planner.rs

```diff
@@ -1409,58 +1581,148 @@ impl PromPlanner {
 
     /// Try to build a DataFusion Literal Expression from PromQL Expr, return
     /// `None` if the input is not a literal expression.
-    fn try_build_literal_expr(expr: &PromExpr) -> Option<DfExpr> {
-        match expr {
+    #[async_recursion]
+    async fn try_build_literal_expr(&mut self, expr: &PromExpr) -> Result<Option<LiteralExpr>> {
+        let output = match expr {
             PromExpr::NumberLiteral(NumberLiteral { val }) => {
                 let scalar_value = ScalarValue::Float64(Some(*val));
-                Some(DfExpr::Literal(scalar_value))
+                Some(LiteralExpr::Const(DfExpr::Literal(scalar_value)))
             }
             PromExpr::StringLiteral(StringLiteral { val }) => {
                 let scalar_value = ScalarValue::Utf8(Some(val.to_string()));
-                Some(DfExpr::Literal(scalar_value))
+                Some(LiteralExpr::Const(DfExpr::Literal(scalar_value)))
             }
             PromExpr::VectorSelector(_)
             | PromExpr::MatrixSelector(_)
             | PromExpr::Extension(_)
             | PromExpr::Aggregate(_)
             | PromExpr::Subquery(_) => None,
-            PromExpr::Call(Call { func, .. }) => {
+            PromExpr::Call(Call { func, args }) => {
                 if func.name == SPECIAL_TIME_FUNCTION {
```

Comment:
What about using `match func.name` to re-organize this `if-else` block? It would be clearer.